File: RxTango/python/examples/fluent_client.py
"""TangoClient fluent builder — one-liners for common patterns.

Demonstrates the TangoClient API: read, map, write, execute.

Mirrors `FluentClient.java`.

Usage:
    python fluent_client.py [device]

    device  defaults to tango://localhost:10000/sys/tg_test/1
"""

import asyncio
import sys
from pathlib import Path

sys.path.insert(0, str(Path(__file__).parent.parent / "src"))

from reactivex.scheduler.eventloop import AsyncIOScheduler

from rxtango import TangoClient


async def main() -> None:
    device = sys.argv[1] if len(sys.argv) > 1 else "tango://localhost:10000/sys/tg_test/1"

    loop      = asyncio.get_running_loop()
    scheduler = AsyncIOScheduler(loop)

    # ── Demo 1: read → map → write ────────────────────────────────────────
    print("Demo 1: read → negate → write\n")
    done1 = asyncio.Event()

    TangoClient() \
        .read(device, "double_scalar") \
        .write(device, "double_scalar_w", value=lambda v: -v) \
        .subscribe(
            on_next=lambda v: print(f"  written: {v:+.6f}"),
            on_completed=lambda: (print("  done.\n"), done1.set()),
            on_error=lambda e: (print(f"  ERROR: {e}"), done1.set()),
            scheduler=scheduler,
        )

    await asyncio.wait_for(done1.wait(), timeout=5.0)

    # ── Demo 2: execute command → read result ─────────────────────────────
    print("Demo 2: execute State command → print\n")
    done2 = asyncio.Event()

    TangoClient() \
        .execute(device, "State") \
        .subscribe(
            on_next=lambda v: print(f"  State = {v}"),
            on_completed=lambda: (print("  done.\n"), done2.set()),
            on_error=lambda e: (print(f"  ERROR: {e}"), done2.set()),
            scheduler=scheduler,
        )

    await asyncio.wait_for(done2.wait(), timeout=5.0)

    # ── Demo 3: multi-step chain ───────────────────────────────────────────
    print("Demo 3: read → calibrate → write → read-back\n")
    done3 = asyncio.Event()

    TangoClient() \
        .read(device, "double_scalar") \
        .map(lambda v: abs(v) * 2.0 + 1.5) \
        .write(device, "double_scalar_w") \
        .read(device, "double_scalar_w") \
        .subscribe(
            on_next=lambda v: print(f"  read-back: {v:+.6f}"),
            on_completed=lambda: (print("  done."), done3.set()),
            on_error=lambda e: (print(f"  ERROR: {e}"), done3.set()),
            scheduler=scheduler,
        )

    await asyncio.wait_for(done3.wait(), timeout=5.0)


if __name__ == "__main__":
    asyncio.run(main())

File: RxTango/cpp/examples/fluent_client.cpp
/**
 * TangoClient fluent builder showcase — all builder features in one example.
 *
 * Demonstrates: read → map (calibrate) → write → execute → map → subscribe.
 *
 * Mirrors Python's fluent_client.py and Java's FluentClient.java.
 *
 * Usage:
 *   ./fluent_client [device]
 *   defaults: tango://localhost:10000/sys/tg_test/1
 */

#include <any>
#include <condition_variable>
#include <cmath>
#include <iostream>
#include <mutex>
#include <string>

#include <rxtango/rxtango.hpp>

int main(int argc, char* argv[]) {
    const std::string device = argc > 1 ? argv[1] : "tango://localhost:10000/sys/tg_test/1";

    std::cout << "TangoClient showcase on " << device << "\n\n";

    std::mutex              m;
    std::condition_variable cv;
    bool                    done = false;

    rxtango::TangoClient()
        .read(device, "double_scalar")                // 1. read
        .map([](std::any v) -> std::any {             // 2. pure calibration transform
            double d = std::any_cast<double>(v);
            std::cout << "  read: " << d << "\n";
            return std::abs(d) * 2.0 + 1.5;
        })
        .write(device, "double_scalar_w")             // 3. write (passes written value through)
        .map([](std::any v) -> std::any {
            std::cout << "  wrote: " << std::any_cast<double>(v) << "\n";
            return v;
        })
        .execute(device, "DevDouble",                 // 4. command with value from previous step
            [](double prev) { return prev; })
        .map([](std::any v) -> std::any {
            std::cout << "  DevDouble result: " << std::any_cast<double>(v) << "\n";
            return v;
        })
        .subscribe(
            [](std::any v) {
                std::cout << "  final: " << std::any_cast<double>(v) << "\n";
            },
            [&](std::exception_ptr e) {
                try { std::rethrow_exception(e); }
                catch (std::exception& ex) { std::cerr << "  ERROR: " << ex.what() << "\n"; }
                std::lock_guard<std::mutex> lk(m); done = true; cv.notify_one();
            },
            [&]() {
                std::cout << "  Done.\n";
                std::lock_guard<std::mutex> lk(m); done = true; cv.notify_one();
            }
        );

    std::unique_lock<std::mutex> lk(m);
    cv.wait_for(lk, std::chrono::seconds(10), [&]{ return done; });
    return 0;
}
